Messaging/EventMessagingHandler.cs

```diff
@@ -0,0 +1,229 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License"); 
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using QuantConnect.Interfaces;
+using QuantConnect.Logging;
+using QuantConnect.Notifications;
+using QuantConnect.Packets;
+
+namespace QuantConnect.Messaging
+{
+    /// <summary>
+    /// Desktop implementation of messaging system for Lean Engine
+    /// </summary>
+    public class EventMessagingHandler : IMessagingHandler
+    {
+        private AlgorithmNodePacket _job;
+        private volatile bool _loaded;
+        private Queue<Packet> _queue; 
+
+        /// <summary>
+        /// Gets or sets whether this messaging handler has any current subscribers.
+        /// When set to false, messages won't be sent.
+        /// </summary>
+        public bool HasSubscribers
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// Initialize the Messaging System Plugin. 
+        /// </summary>
+        public void Initialize()
+        {
+            _queue = new Queue<Packet>();
+
+            ConsumerReadyEvent += () => { _loaded = true; };
+        }
+
+        public void LoadingComplete()
+        {
+            _loaded = true;
+        }
+
+        /// <summary>
+        /// Set the user communication channel
+        /// </summary>
+        /// <param name="job"></param>
+        public void SetAuthentication(AlgorithmNodePacket job)
+        {
+            _job = job;
+        }
+
+        public delegate void DebugEventRaised(DebugPacket packet);
+        public event DebugEventRaised DebugEvent;
+
+        public delegate void LogEventRaised(LogPacket packet);
+        public event LogEventRaised LogEvent;
+
+        public delegate void RuntimeErrorEventRaised(RuntimeErrorPacket packet);
+        public event RuntimeErrorEventRaised RuntimeErrorEvent;
+
+        public delegate void HandledErrorEventRaised(HandledErrorPacket packet);
+        public event HandledErrorEventRaised HandledErrorEvent;
+
+        public delegate void BacktestResultEventRaised(BacktestResultPacket packet);
+        public event BacktestResultEventRaised BacktestResultEvent;
+
+        public delegate void ConsumerReadyEventRaised();
+        public event ConsumerReadyEventRaised ConsumerReadyEvent;
```

Comment:
I envisioned this all being one event handler of type packet... I think that type of pattern will lead to less maintenance